### PR TITLE
Add signal for personal data

### PR DESCRIPTION
**Pull request by vtran99**
Request to add a new signal to get extra data for the user personal data profile's page.

### CODE REVIEW

File: indico/modules/users/controllers.py [hunk 1]
```diff
@@ -202,19 +202,24 @@ def _check_access(self):
...
             current_affiliation = AffiliationSchema().dump(self.user.affiliation_link)
         has_predefined_affiliations = Affiliation.query.filter(~Affiliation.is_deleted).has_rows()
+        extra_data = values_from_signal(signals.users.personal_data.send(self.user), as_list=True)
         return WPUserPersonalData.render_template('personal_data.html', 'personal_data', user=self.user,
                                                   titles=titles, user_values=user_values, locked_fields=locked_fields,
                                                   locked_field_message=multipass.locked_field_message,
                                                   current_affiliation=current_affiliation,
-                                                  has_predefined_affiliations=has_predefined_affiliations)
+                                                  has_predefined_affiliations=has_predefined_affiliations,
+                                                  extra_data=extra_data[0] if extra_data else None)
```
Comment (ThiefMaster) [inline indico/modules/users/controllers.py:222]: this breaks if more than one plugin uses this signal. you may want to merge the data.

Comment (vtran99) [reply indico/modules/users/controllers.py:222]: I added the context when sending signal, so should not have multiple uses for the same class.

Comment (ThiefMaster) [reply indico/modules/users/controllers.py:222]: i was thinking about cases where one has two plugins that both make use of the signal.. and more context does not help there

Comment (vtran99) [reply indico/modules/users/controllers.py:222]: Sorry, I still can't see the case you mentioned:
I expect the one using the signal to check if sender if RHPersonalData, then in the same Indico implementation, it means that connect is to do sth in the PD page,
then how the same Indico implementation can have another plugin working on that page without checking a signal connect already exists ?

Nevertheless, I will change code following your suggestion: chaining results and remove RH.

Comment (ThiefMaster) [reply indico/modules/users/controllers.py:222]: My point is that Indico has no control over what plugins someone installs.

Let's say someone installs a "gender" plugin and someone else a "favorite color" plugin. So now you have two plugins, one of them returning `{'gender': 'male'}` and the other is returning `{'favorite_color': 'yellow'}`.

Comment (vtran99) [reply indico/modules/users/controllers.py]: Done as suggested: merge data and signal renamed to "extra_personal_data".

File: indico/modules/users/controllers.py [hunk 2]
```diff
@@ -213,13 +213,15 @@ def _process(self):
...
+        extra_data = {}
+        for data in values_from_signal(signals.users.extra_personal_data.send(self.user), as_list=True):
+            extra_data.update(data)
         return WPUserPersonalData.render_template('personal_data.html', 'personal_data', user=self.user,
                                                   titles=titles, user_values=user_values, locked_fields=locked_fields,
                                                   locked_field_message=multipass.locked_field_message,
                                                   current_affiliation=current_affiliation,
                                                   has_predefined_affiliations=has_predefined_affiliations,
-                                                  extra_data=extra_data[0] if extra_data else None)
+                                                  extra_data=extra_data if extra_data else None)
```
Comment (ThiefMaster) [inline indico/modules/users/controllers.py:224]:
```suggestion
                                                  extra_data=extra_data)
```

it's always a dict now

Comment (ThiefMaster) [reply indico/modules/users/controllers.py:224]: or do you want it to be None if it's empty? `extra_data=(extra_data or None)`

but maybe an empty dict is nicer?

Comment (vtran99) [reply indico/modules/users/controllers.py]: Done.

**Comment by ThiefMaster** [comment]
I'm fine merging it as it is now, but just wondering: Shouldn't the extraData also be included in the `userValues`, so custom form fields can properly get the value and you have all the final-form logic (enabling the submit button once some field changed etc)?

**Comment by vtran99** [comment]
> I'm fine merging it as it is now, but just wondering: Shouldn't the extraData also be included in the `userValues`, so custom form fields can properly get the value and you have all the final-form logic (enabling the submit button once some field changed etc)?

In fact, we use extraDta just for display (list of options for drop-down list fields) and we patched the UserPersonalDataSchema to add the fields to be updated, so it works for UN.
As for your suggestion, yes that would be more complete but then how can I make the schema of extraData flexible so plugins can define whatever types fields: 
is there a plugin mechanism to add fileds in schema ?

**Comment by ThiefMaster** [comment]
Yes, there's a `schema_post_dump` signal you can use. It runs in an `@post_dump` marshmallow hook for any schema. In there you could use a custom schema to dump your extra data, and add it to the data coming from the original schema.

And it sounds like you already have something that does everything properly when combining the two (this signal + your patching / the schema_post_dump signal).

**Comment by vtran99** [comment]
> Yes, there's a `schema_post_dump` signal you can use. It runs in an `@post_dump` marshmallow hook for any schema. In there you could use a custom schema to dump your extra data, and add it to the data coming from the original schema.
> 
> And it sounds like you already have something that does everything properly when combining the two (this signal + your patching / the schema_post_dump signal). To confirm this I'd just recommend you to test if the form field data tracking works correctly, ie that: the submit button becomes enabled when you change the text in one of your custom fields, and that it becomes disabled again if you change it back to the initial value.

I don't need the signal anymore as the post_dump is enough to add the extra data for display purpose, and we patch the original class so the update mechanism and the form data tracking works.
The signal could be useful if there was a way to change also the schema when updating, that would allow other Indico implementations without patching mechanism to add data to display and update.

**Comment by vtran99** [comment]
As the new signal is not necessary anymore, please ignore this PR.
I created a new one (https://github.com/indico/indico/pull/6117) to get the user titles interceptable.
Thanks,

**Pull request closed without merging**